src/__tests__/render.test.tsx

```diff
@@ -285,3 +281,160 @@ test('RenderAPI type', () => {
   render(<Banana />) as RenderAPI;
   expect(true).toBeTruthy();
 });
+
+test('should throw when rendering a string outside a text component', () => {
+  expect(() =>
+    render(<View>hello</View>, {
+      unstable_validateStringsRenderedInText: true,
+    })
+  ).toThrowErrorMatchingInlineSnapshot(
+    `"Text strings must be rendered within a host Text component. Text hello was rendered in a View"`
+  );
+});
+
+const profilerErrorMessage =
+  'The above error occurred in the <Profiler> component';
+
+test('should throw an error when rerendering with text outside of Text component', () => {
+  // eslint-disable-next-line no-console
+  console.error = (errorMessage: string) => {
+    if (!errorMessage.includes(profilerErrorMessage)) {
+      originalConsoleError(errorMessage);
+    }
+  };
+  const { rerender } = render(<View />, {
+    unstable_validateStringsRenderedInText: true,
+  });
+
+  expect(() => rerender(<View>hello</View>)).toThrowError(
+    'Text strings must be rendered within a host Text component.'
+  );
+  // eslint-disable-next-line no-console
+  console.error = originalConsoleError;
+});
+
+const ErrorComponent = () => {
+  const [shouldDisplayText, setShouldDisplayText] = React.useState(false);
+
+  if (!shouldDisplayText) {
+    return (
+      <Pressable
+        onPress={() => {
+          setShouldDisplayText(true);
+        }}
+      >
+        <Text>Display text</Text>
+      </Pressable>
+    );
+  }
+
+  return <View>text rendered outside text component</View>;
+};
+
+test('should throw an error when strings are rendered outside Text', () => {
+  // eslint-disable-next-line no-console
+  console.error = (errorMessage: string) => {
+    if (!errorMessage.includes(profilerErrorMessage)) {
+      originalConsoleError(errorMessage);
+    }
+  };
+  const { getByText } = render(<ErrorComponent />, {
+    unstable_validateStringsRenderedInText: true,
+  });
+
+  expect(() => fireEvent.press(getByText('Display text'))).toThrowError(
+    'Text strings must be rendered within a host Text component.'
+  );
+  // eslint-disable-next-line no-console
+  console.error = originalConsoleError;
+});
+
+test('it should not throw for texts nested in fragments', () => {
+  expect(() =>
+    render(
+      <Text>
+        <>hello</>
+      </Text>,
+      { unstable_validateStringsRenderedInText: true }
+    )
+  ).not.toThrow();
+});
+
+test('it should not throw if option validateRenderedString is false', () => {
+  expect(() => render(<View>hello</View>)).not.toThrow();
+});
+
+test(`it should throw 
+      - when one of the children is a text
+      - and the parent is not a Text component`, () => {
+  expect(() =>
+    render(
+      <View>
+        <Text>hello</Text>
+        hello
+      </View>,
+      { unstable_validateStringsRenderedInText: true }
+    )
+  ).toThrowErrorMatchingInlineSnapshot(
+    `"Text strings must be rendered within a host Text component. Text hello was rendered in a View"`
```

Comment:
Lovely error message.

I'm curious btw, what does the stacktrace looks like?

Reply:
The stack trace looks like this :

<img width="980" alt="Capture d’écran 2022-09-13 à 20 14 37" src="https://user-images.githubusercontent.com/64224599/189979455-226ec7cf-9586-4b21-9134-cb7837c17fa9.png">

And when the error occurs during a rerender you get this additional warning that an error occurred in the React Profiler

<img width="681" alt="Capture d’écran 2022-09-13 à 20 18 39" src="https://user-images.githubusercontent.com/64224599/189979933-960c7ffa-7c4b-480f-afd5-c2bab65ee3a3.png">

Reply:
Thanks!